### PR TITLE
Interpreter implementation of batch norm bprop

### PR DESCRIPTION
**Pull request by varun-intel**
-> Implements batch norm bprop in the interpreter backend and adds a test to verify the implementation.

### CODE REVIEW

File: src/ngraph/op/batch_norm.cpp [hunk 1]
```diff
@@ -269,13 +271,31 @@ void ngraph::op::BatchNorm::generate_adjoints(autodiff::Adjoints& adjoints,
     std::vector<std::shared_ptr<Node>> goes(get_outputs().size());
     if (this->get_training_flag() && get_input_size() == 3)
     {
+        std::set<size_t> output_input_found;
```
Comment (diyessi) [inline src/ngraph/op/batch_norm.cpp]: @varun-intel There needs to be a comment describing what this is all about.
@Krovatkin Whatever this is doing, it's way too complicated if it needs to be done this way.

File: src/ngraph/runtime/reference/batch_norm.hpp [hunk 2]
```diff
@@ -85,13 +92,38 @@ namespace ngraph
...
+            void batch_norm_three_outputs(double eps,
+                                          const T* arg0,
+                                          const T* arg1,
+                                          const T* arg2,
+                                          T* out0,
+                                          T* out1,
+                                          T* out2,
+                                          const Shape& arg2_shape)
+            {
+                std::unique_ptr<T[]> centered(new T[shape_size(arg2_shape)]());
```
[the remark this answers is not in the document]
Comment (diyessi) [reply src/ngraph/runtime/reference/batch_norm.hpp]: It's a dynamically created array since the number of elements isn't known at compile-time; the `unique_ptr` ensures it gets deleted when `batch_norm_three_outputs_with_intermediates` returns.

[the remark this answers is not in the document]
Comment (Krovatkin) [reply src/ngraph/runtime/reference/batch_norm.hpp]: @diyessi are we sure it does this in this *particular* case? because the very next thing we do with those `unique_ptr`s is to call `get` on them which makes those `unique_ptr`s to release the ownership over those arrays, so at the point when the unique_ptrs get destructed, they don't own any storage anymore.
I'd suggest using `std::vector<T>` and `data()` to get the underlying storage. Since vectors' lifetime covers intermediate arrays lifetime and yet the vectors would get destroyed after a function is done, this would give us semi-robust memory management w/o needing to call `delete` explicitly

File: src/ngraph/op/batch_norm.cpp (hunk 1 above)
Comment (Krovatkin) [reply src/ngraph/op/batch_norm.cpp]: this is due to a fact that users aren't ordered but this code needs to be able to tell which between output, mean or variance, so we need to recover that order. I'm going to add a simple helper that will hopefully make it less painful.... but there isn't a very good way around this problem.

Comment (Krovatkin) [reply src/ngraph/op/batch_norm.cpp]: @diyessi @varun-intel with #1154 
we could do something like this : 
```cpp

//TODO: define get_goe_or_default to create a new GOE if one doesn't exist already
//We could define `get_goe_or_default` in `BatchNorm` class, then we wouldn't need the second argument.

auto goes = op::GetOutputElement::get_get_output_elements(this->shared_from_this());
auto output = get_goe_or_default(goes.at(0), this->shared_from_this());
auto mean = get_goe_or_default(goes.at(1), this->shared_from_this());
auto var = get_goe_or_default(goes.at(2), this->shared_from_this());

```

File: src/ngraph/runtime/reference/batch_norm.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (rkimballn1) [reply src/ngraph/runtime/reference/batch_norm.hpp]: Just a drive-by comment. `get` on a `unique_ptr` does not make it release ownership of it's data, it only returns a pointer to it's data. The `unique_ptr` is not destructed by calling `get`

**Comment by varun-intel** [comment]
TESTNOW

**Comment by aprocter** [comment]
:confetti_ball: :tada:
